cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -268,35 +267,79 @@ protected void triggerShutdown() {
    */
-  private void persistMetricsMessageIds(Deque<MetricValues> metricValues, Map<TopicIdMetaKey, byte[]> messageIds) {
+  private void persistMetricsMessageIds(Deque<MetricValues> metricValues,
+                                        Map<TopicIdMetaKey, MetricsProcessorStats> messageIds) {
     try {
+      // we update the processing stats in meta table before we persist metrics,
+      // in case of huge-delay in writing to metrics table (maybe due to region server issues),
+      // this will help surface the last processed timestamp information to the users.
+      updateProcessingStats(messageIds);
       if (!metricValues.isEmpty()) {
-        persistMetrics(metricValues);
+        persistMetrics(metricValues, messageIds);
       }
       persistMessageIds(messageIds);
+      // reset processing stats, so the latest processing stats will get updated in the next iteration.
+      resetProcessingStats(messageIds);
     } catch (Exception e) {
       LOG.warn("Failed to persist metrics.", e);
     }
   }
 
+  private void updateProcessingStats(Map<TopicIdMetaKey, MetricsProcessorStats> messageIds) {
+    try {
+      // messageIds can be empty if the current thread fetches nothing while other threads keep fetching new metrics
+      // and haven't updated messageId's of the corresponding topics
+      if (!messageIds.isEmpty()) {
+        metaTable.saveMetricsProcessorStats(messageIds);
+      }
+    } catch (Exception e) {
+      LOG.warn("Failed to update processing stats of consumed messages.", e);
+    }
+  }
+
+  private void resetProcessingStats(Map<TopicIdMetaKey, MetricsProcessorStats> messageIds) {
+    for (MetricsProcessorStats metaInfo : messageIds.values()) {
+      metaInfo.resetMetaInfo();
+    }
+  }
+
   /**
    * Persist metrics into metric store
    *
    * @param metricValues a non-empty deque of {@link MetricValues}
    */
-  private void persistMetrics(Deque<MetricValues> metricValues) throws Exception {
+  private void persistMetrics(Deque<MetricValues> metricValues,
+                              Map<TopicIdMetaKey, MetricsProcessorStats> topicMetrics) throws Exception {
     long now = System.currentTimeMillis();
     long lastMetricTime = metricValues.peekLast().getTimestamp();
-    long delay = now - TimeUnit.SECONDS.toMillis(lastMetricTime);
-    metricValues.add(
-      new MetricValues(metricsContextMap, TimeUnit.MILLISECONDS.toSeconds(now),
-                       ImmutableList.of(
-                         new MetricValue(processMetricName, MetricType.COUNTER, metricValues.size()),
-                         new MetricValue(delayMetricName, MetricType.GAUGE, delay))));
+    List<MetricValue> topicLevelDelays = new ArrayList<>();
+
+    //add topic level delay metrics
+    for (Map.Entry<TopicIdMetaKey, MetricsProcessorStats> entry : topicMetrics.entrySet()) {
+      if (entry.getValue().getOldestMetricsTimestamp() != null) {
+        String metricName = String.format("%s.topic.%s.oldest.%s",
```

Comment:
then you can create the Map once at startup, and have these metric names in the MetricsProcessorStats. It is really a waste to re format this same string every time.